File: backend/src/services/tools/text.ts
/**
 * Text Utilities Tool — word/character counting, case conversion, encoding, etc.
 *
 * AI detects when the user wants text analysis or transformation.
 */

import { registerTool } from './index';
import type { ToolDefinition, ToolExecutor } from './index';

// ─── Helper functions ─────────────────────────────────────

function wordCount(text: string): number {
  const words = text.trim().split(/\s+/).filter(Boolean);
  return words.length;
}

function charCount(text: string): number {
  return text.length;
}

function charCountNoSpaces(text: string): number {
  return text.replace(/\s/g, '').length;
}

function lineCount(text: string): number {
  return text.split('\n').length;
}

function sentenceCount(text: string): number {
  const sentences = text.split(/[.!?]+/).filter((s) => s.trim().length > 0);
  return sentences.length;
}

function reverseString(text: string): string {
  return text.split('').reverse().join('');
}

function toUpperCase(text: string): string {
  return text.toUpperCase();
}

function toLowerCase(text: string): string {
  return text.toLowerCase();
}

function toTitleCase(text: string): string {
  return text.replace(/\w\S*/g, (word) => {
    return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
  });
}

function toCamelCase(text: string): string {
  return text
    .replace(/[-_\s]+(.)/g, (_, c) => c.toUpperCase())
    .replace(/^[A-Z]/, (c) => c.toLowerCase())
    .replace(/[-_\s]/g, '');
}

function toSnakeCase(text: string): string {
  return text
    .replace(/([A-Z])/g, '_$1')
    .replace(/[-_\s]+/g, '_')
    .toLowerCase()
    .replace(/^_/, '');
}

function base64Encode(text: string): string {
  return Buffer.from(text, 'utf-8').toString('base64');
}

function base64Decode(text: string): string {
  return Buffer.from(text, 'base64').toString('utf-8');
}

function urlEncode(text: string): string {
  return encodeURIComponent(text);
}

function urlDecode(text: string): string {
  return decodeURIComponent(text);
}

function countVowels(text: string): number {
  return (text.match(/[aeiouáéíóúàèìòùäëïöüâêîôû]/gi) || []).length;
}

function countConsonants(text: string): number {
  return (text.match(/[bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ]/g) || []).length;
}

const OPERATIONS: Record<string, (text: string) => string> = {
  'uppercase': toUpperCase,
  'lowercase': toLowerCase,
  'titlecase': toTitleCase,
  'camelcase': toCamelCase,
  'snakecase': toSnakeCase,
  'reverse': reverseString,
  'base64encode': base64Encode,
  'base64decode': base64Decode,
  'urlencode': urlEncode,
  'urldecode': urlDecode,
};

function detectOperation(input: string): string | null {
  const lower = input.toLowerCase();
  if (lower.includes('uppercase') || lower.includes('upper case') || lower.includes('capitalize') || lower.includes('all caps')) return 'uppercase';
  if (lower.includes('lowercase') || lower.includes('lower case')) return 'lowercase';
  if (lower.includes('title case') || lower.includes('titlecase')) return 'titlecase';
  if (lower.includes('camel case') || lower.includes('camelcase')) return 'camelcase';
  if (lower.includes('snake case') || lower.includes('snakecase')) return 'snakecase';
  if (lower.includes('reverse') || lower.includes('backwards')) return 'reverse';
  if (lower.includes('base64 encode') || lower.includes('base64 encode')) return 'base64encode';
  if (lower.includes('base64 decode') || lower.includes('decode base64')) return 'base64decode';
  if (lower.includes('url encode')) return 'urlencode';
  if (lower.includes('url decode')) return 'urldecode';
  return null;
}

// ─── Tool Definition ──────────────────────────────────────

const definition: ToolDefinition = {
  id: 'text',
  name: 'Text Utilities',
  description: 'Analyze and transform text: word/char/line/sentence count, case conversion, encoding (base64, URL), and more',
  version: '1.0.0',
  icon: '📝',
  params: [
    { name: 'operation', type: 'string', description: 'What to do: count, uppercase, lowercase, titlecase, camelcase, snakecase, reverse, base64encode, base64decode, urlencode, urldecode', required: true },
    { name: 'text', type: 'string', description: 'The text to analyze or transform', required: true },
  ],
};

const execute: ToolExecutor = async (params, ctx) => {
  const text = String(params.text || params.query || ctx.userInput || '');
  const operation = String(params.operation || detectOperation(ctx.userInput) || 'count').toLowerCase();

  if (!text.trim()) {
    return { success: false, output: 'Please provide some text to analyze or transform.' };
  }

  // Analysis operations
  if (operation === 'count' || operation === 'stats' || operation === 'analyze') {
    const wc = wordCount(text);
    const cc = charCount(text);
    const ccns = charCountNoSpaces(text);
    const lc = lineCount(text);
    const sc = sentenceCount(text);
    const vowels = countVowels(text);
    const consonants = countConsonants(text);

    return {
      success: true,
      output: `📊 Text Statistics:\n• Words: ${wc}\n• Characters: ${cc} (${ccns} without spaces)\n• Lines: ${lc}\n• Sentences: ${sc}\n• Vowels: ${vowels}\n• Consonants: ${consonants}`,
      data: { words: wc, chars: cc, charsNoSpaces: ccns, lines: lc, sentences: sc, vowels, consonants },
    };
  }

  // Transformation operations
  const transformer = OPERATIONS[operation];
  if (transformer) {
    try {
      const result = transformer(text);
      return {
        success: true,
        output: result,
        data: { operation, input: text, output: result },
      };
    } catch (e) {
      return {
        success: false,
        output: `Failed to ${operation}: ${e instanceof Error ? e.message : String(e)}`,
      };
    }
  }

  return {
    success: false,
    output: `Unknown operation "${operation}". Available: count/stats, uppercase, lowercase, titlecase, camelcase, snakecase, reverse, base64encode, base64decode, urlencode, urldecode`,
  };
};

// Auto-detect text utility intent
function detect(input: string): { confidence: number; params: Record<string, unknown> } | null {
  const lower = input.toLowerCase();

  const isAnalysis =
    lower.includes('word count') || lower.includes('character count') ||
    lower.includes('how many words') || lower.includes('how many characters') ||
    lower.includes('count words') || lower.includes('count characters') ||
    lower.includes('text stats') || lower.includes('analyze this text');

  const isTransform =
    lower.includes('uppercase') || lower.includes('lowercase') ||
    lower.includes('title case') || lower.includes('camel case') ||
    lower.includes('snake case') || lower.includes('reverse') ||
    lower.includes('base64') || lower.includes('url encode') ||
    lower.includes('url decode');

  if (isAnalysis || isTransform) {
    return { confidence: 0.7, params: { operation: isAnalysis ? 'count' : detectOperation(input) || 'count', query: input } };
  }

  return null;
}

// ─── Register ─────────────────────────────────────────────

export function registerTextTool(): void {
  registerTool(definition, execute);
  console.log('[tools] Text Utilities registered');
}

export { detectOperation, wordCount, charCount, lineCount };

File: backend/app/tools/text.py
"""Text Utilities Tool — mirrors backend/src/services/tools/text.ts."""

from __future__ import annotations

import base64
import re
import urllib.parse
from typing import Any, Callable

from ..logger import info as log_info
from . import ToolContext, ToolDefinition, ToolParam, ToolResult, register_tool


def word_count(text: str) -> int:
    return len([w for w in re.split(r"\s+", text.strip()) if w])


def char_count(text: str) -> int:
    return len(text)


def line_count(text: str) -> int:
    return len(text.split("\n"))


def to_title_case(text: str) -> str:
    return re.sub(
        r"\w\S*", lambda m: m.group(0)[0].upper() + m.group(0)[1:].lower(), text
    )


def to_camel_case(text: str) -> str:
    out = re.sub(r"[-_\s]+(.)", lambda m: m.group(1).upper(), text)
    if out and out[0].isupper():
        out = out[0].lower() + out[1:]
    return re.sub(r"[-_\s]", "", out)


def to_snake_case(text: str) -> str:
    out = re.sub(r"([A-Z])", r"_\1", text)
    out = re.sub(r"[-_\s]+", "_", out).lower()
    return out[1:] if out.startswith("_") else out


OPERATIONS: dict[str, Callable[[str], str]] = {
    "uppercase": str.upper,
    "lowercase": str.lower,
    "titlecase": to_title_case,
    "camelcase": to_camel_case,
    "snakecase": to_snake_case,
    "reverse": lambda t: t[::-1],
    "base64encode": lambda t: base64.b64encode(t.encode("utf-8")).decode("ascii"),
    "base64decode": lambda t: base64.b64decode(t).decode("utf-8", "replace"),
    "urlencode": lambda t: urllib.parse.quote(t, safe=""),
    "urldecode": lambda t: urllib.parse.unquote(t),
}


def detect_operation(input_text: str) -> str | None:
    lower = input_text.lower()
    if any(k in lower for k in ("uppercase", "upper case", "capitalize", "all caps")):
        return "uppercase"
    if any(k in lower for k in ("lowercase", "lower case")):
        return "lowercase"
    if any(k in lower for k in ("title case", "titlecase")):
        return "titlecase"
    if any(k in lower for k in ("camel case", "camelcase")):
        return "camelcase"
    if any(k in lower for k in ("snake case", "snakecase")):
        return "snakecase"
    if "reverse" in lower or "backwards" in lower:
        return "reverse"
    if "base64 encode" in lower:
        return "base64encode"
    if "base64 decode" in lower or "decode base64" in lower:
        return "base64decode"
    if "url encode" in lower:
        return "urlencode"
    if "url decode" in lower:
        return "urldecode"
    return None


DEFINITION = ToolDefinition(
    id="text",
    name="Text Utilities",
    description=(
        "Analyze and transform text: word/char/line/sentence count, case conversion, "
        "encoding (base64, URL), and more"
    ),
    version="1.0.0",
    icon="\U0001f4dd",
    params=[
        ToolParam(
            name="operation",
            type="string",
            description=(
                "What to do: count, uppercase, lowercase, titlecase, camelcase, snakecase, "
                "reverse, base64encode, base64decode, urlencode, urldecode"
            ),
            required=True,
        ),
        ToolParam(name="text", type="string", description="The text to analyze or transform", required=True),
    ],
)


async def execute(params: dict[str, Any], ctx: ToolContext) -> ToolResult:
    text = str(params.get("text") or params.get("query") or ctx.userInput or "")
    operation = str(
        params.get("operation") or detect_operation(ctx.userInput) or "count"
    ).lower()

    if not text.strip():
        return ToolResult(success=False, output="Please provide some text to analyze or transform.")

    if operation in ("count", "stats", "analyze"):
        wc = word_count(text)
        cc = char_count(text)
        ccns = len(re.sub(r"\s", "", text))
        lc = line_count(text)
        sc = len([s for s in re.split(r"[.!?]+", text) if s.strip()])
        vowels = len(re.findall(r"[aeiou\u00e1\u00e9\u00ed\u00f3\u00fa\u00e0\u00e8\u00ec\u00f2\u00f9\u00e4\u00eb\u00ef\u00f6\u00fc\u00e2\u00ea\u00ee\u00f4\u00fb]", text, re.IGNORECASE))
        consonants = len(re.findall(r"[bcdfghjklmnpqrstvwxyz]", text, re.IGNORECASE))
        return ToolResult(
            success=True,
            output=(
                f"\U0001f4ca Text Statistics:\n\u2022 Words: {wc}\n\u2022 Characters: {cc} ({ccns} without spaces)"
                f"\n\u2022 Lines: {lc}\n\u2022 Sentences: {sc}\n\u2022 Vowels: {vowels}\n\u2022 Consonants: {consonants}"
            ),
            data={
                "words": wc, "chars": cc, "charsNoSpaces": ccns, "lines": lc,
                "sentences": sc, "vowels": vowels, "consonants": consonants,
            },
        )

    transformer = OPERATIONS.get(operation)
    if transformer:
        try:
            result = transformer(text)
            return ToolResult(
                success=True, output=result, data={"operation": operation, "input": text, "output": result}
            )
        except Exception as e:  # noqa: BLE001
            return ToolResult(success=False, output=f"Failed to {operation}: {e}")

    return ToolResult(
        success=False,
        output=(
            f'Unknown operation "{operation}". Available: count/stats, uppercase, lowercase, titlecase, '
            "camelcase, snakecase, reverse, base64encode, base64decode, urlencode, urldecode"
        ),
    )


def register_text_tool() -> None:
    register_tool(DEFINITION, execute)
    log_info("[tools] Text Utilities registered")
